### DISCUSSION TITLE
Content rendered via HTML doesn't print on some printers

### DISCUSSION THREAD

**Discussion by ZachSARAO** [pymupdf/PyMuPDF#3242] [Q&A]
### Description of the bug

When printing a PDF rendered via `PyMuPDF`, I have found that physical printouts omit text in some cases:

```py
insert_htmlbox(
    some_rectangle,
    f'<div style="font-family: Helvetica; font-weight: bold; text-align: center; font-size: 8px; line-height: {height}px; color: #587BBD;"><b>{title}</b></div>',
)
```

I can see that content on PDF on screen, and printing out on my printer that text is visible. However, a different printer omitted this content.

Please let me know if I'm misusing the library in any way / or in general what could cause this?

### How to reproduce the bug

A PDF produced via PyMuPDF prints correctly on some printers, but not on others.

### PyMuPDF version

1.23.9

### Operating system

Linux

### Python version

3.10

**Comment by JorjMcKie** [comment]
This is obviously a question and no bug report.
Transferring to "Discussions".

**Comment by JorjMcKie** [comment]
Please provide an example of non-printable text.
Otherwise, if the file looks good in a PDF viewer, then this problem is outside the scope of this library.

**Reply by ZachSARAO** [reply]
The text specified above printed on my printer, but not another printer. Other text using `Page.insert_text` and `Page.insert_textbox` rendered fine.

I am interested to know what could cause this - even if out of the scope of this library - as I would like to avoid sending someone a PDF that doesn't print correctly on their side (without knowing what pinter they use, OS they use, etc).

**Reply by JorjMcKie** [reply]
If you do not specify fonts that you want to be used, then the underlying `Story` object will choose ones it deems appropriate.
Specifying a font for use in `insert_htmlbox` is exemplified for example [here](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-specify-your-own-fonts).

**Reply by JorjMcKie** [reply]
BTW I still see no example text to get an idea about what might go wrong ...

**Reply by ZachSARAO** [reply]
This code results in a PDF that does display on a screen, but does not print on one of the printers I am testing:

```py
doc = fitz.open()
doc.new_page()
doc[0].insert_htmlbox(
    fitz.Rect(100, 100, 400, 400),
    f'<div style="font-family: Helvetica; font-weight: bold; text-align: center; font-size: 8px; line-height: 100px; color: #587BBD;"><b>Some bold text</b></div>',
)
doc.save("./test.pdf")
```

**Reply by ZachSARAO** [reply]
The metadata looks like this:

```
<<
  /Type /Page
  /MediaBox [ 0 0 595 842 ]
  /Rotate 0
  /Resources 3 0 R
  /Parent 2 0 R
  /Contents [ 14 0 R ]
>>
```

(I can see that I accidentally set font-weight to bold and specified a `<b>` tag. That is not the problem, as non-bold text also does not print.

It looks like the fonts are not specified in the metadata?

**Reply by ZachSARAO** [reply]
[test-disappearing-text.pdf](https://github.com/pymupdf/PyMuPDF/files/14584279/test-disappearing-text.pdf)

**Reply by JorjMcKie** [reply]
Simply believe me: everything is in order with the produced PDF!
The font and its xref can be found easily like this:
```python
print(page.get_fonts())
[(6, 'cid', 'Type0', 'Nimbus Sans Bold', 'F0', 'Identity-H')]
print(doc.xref_object(6))
<<
  /Type /Font
  /Subtype /Type0
  /BaseFont /Nimbus#20Sans#20Bold
  /Encoding /Identity-H
  /ToUnicode 7 0 R
  /DescendantFonts [ 8 0 R ]
>>
```
The PDF couldn't be more straightforward - a printer incapable of printing this has a problem!

**Reply by ZachSARAO** [reply]
Thank you for having a look. I am still interested in figuring out exactly what that printer problem could be though, as the printer prints everyone else's PDFs fine...

I'll find out the printer model in a bit. Could it be that a printer doesn't include the `Nimbus Sans Bold` font? I'm producing the PDF on Ubuntu 22, and emailing the PDF to a Mac, where it's printed from.

Since I specified Helvetica, why wouldn't the HTML use Helvetica?

**Reply by JorjMcKie** [reply]
A printer does not know fonts! Only the document-producing application does.
The Nimbus font in the PDF is a Helvetica-implementation!
The word "Helvetica" as such is just a definition of details that together describe what is typical for a font to be a Helvetica implementation.

When sending stuff to a printer, then this is a stream of bits that the hardware translates and hopefully understands
No idea what is going wrong in your setup ... e-mail? ... e-mail attachment?
Do all PDFs you are sending over this way show the problem, etc., etc.

**Reply by ZachSARAO** [reply]
I sent a PDF as an email attachment that contained text drawn via html and via the insert_text method. The text written via insert_text shows up fine, whereas the text drawn via html does not (in the same pdf). I've just tested that the email attachment re-sent to me still includes printable text on my side

**Reply by ZachSARAO** [reply]
On a slightly different note, when I choose to print this PDF on a Mac the default printer setting is '100%' whereas with other pdfs, the default printer setting is 'scale to fit'. this means that unless I scale explicitly on a mac, some of the edge to edge content is cut (header and footer).

Looking around the metadata I can't see why that would be the case. Do you have any pointers?

**Reply by JorjMcKie** [reply]
When sending something to a print - depending on the operating system and the printer hardware - you usually make a choice about the output properties like scaling, color and grayscale, how many document pages on a paper sheet, double sided printing and what not.

**Reply by ZachSARAO** [reply]
>I sent a PDF as an email attachment that contained text drawn via html and via the insert_text method. The text written via insert_text shows up fine, whereas the text drawn via html does not (in the same pdf). I've just tested that the email attachment re-sent to me still includes printable text on my side

Sorry - that to confirm that nothing is being stripped from the PDF during email transmission (i.e. I still don't know why at least one printer doesn't print some text)

 >When sending something to a print - depending on the operating system and the printer hardware - you usually make a choice about the output properties like scaling, color and grayscale, how many document pages on a paper sheet, double sided printing and what not.

I have two similar PDFs - one that defaults to 'letter' size, and one that defaults to A4 size despite that the page dimensions are roughly the same (the one that defaults to letter size is slightly different `/MediaBox [ 0 0 595.2756 841.8898 ]`.

What can prompt the printer settings to select one size (or scaling) over another?
